app/src/main/java/com/h/pixeldroid/fragments/CameraFragment.kt

```diff
@@ -1,39 +1,283 @@
+        outputSurfaces.add(reader.surface)
+        outputSurfaces.add(Surface(textureView.surfaceTexture))
+
+        // Handle rotation ?!
+        pictureRequestBuilder.set(
+            CaptureRequest.JPEG_ORIENTATION,
+            getJpegOrientation(
+                manager.getCameraCharacteristics(cameraDevice.id),
+                activity?.windowManager?.defaultDisplay?.rotation!!
+            )
+        )
+
+        cameraDevice.createCaptureSession(outputSurfaces, picturetateCallback , null);
+    }
+
+    private val picturetateCallback = object : CameraCaptureSession.StateCallback() {
+        override fun onConfigured(session: CameraCaptureSession) {
+            try {
+                session.capture(pictureRequestBuilder.build(), pictureCaptureListener, null)
+            } catch (e :CameraAccessException) {
+                e.printStackTrace()
+            }
+        }
+
+        override fun onConfigureFailed(session: CameraCaptureSession) {
+        }
+    }
+
+    private val pictureCaptureListener = object : CameraCaptureSession.CaptureCallback() {
+        override fun onCaptureCompleted(
+            session: CameraCaptureSession,
+            request: CaptureRequest,
+            result: TotalCaptureResult
+        ) {
+            super.onCaptureCompleted(session, request, result)
+        }
+    }
+
+    private val readerListener = OnImageAvailableListener { reader ->
+        lateinit var image : Image
+        try {
+            image = reader.acquireLatestImage()
+            val buffer = image.planes[0].buffer;
+            val bytes = ByteArray(buffer.capacity())
+            buffer.get(bytes);
+
+            val filename = "tmp_capture.png"
+            val file = File.createTempFile(filename, null, context?.cacheDir)
+            file.writeBytes(bytes)
+
+            pictureHandle(file.toUri() as Uri?)
+
+        } catch (e: FileNotFoundException) {
+            e.printStackTrace();
+        } catch (e: IOException) {
+            e.printStackTrace();
+        } finally {
+            image.close()
+        }
+    }
+
+    private fun getJpegOrientation(c: CameraCharacteristics, deviceOrientation: Int): Int {
+        var deviceOrientation = deviceOrientation
+        if (deviceOrientation == ORIENTATION_UNKNOWN) return 0
+        val sensorOrientation = c[CameraCharacteristics.SENSOR_ORIENTATION]!!
+
+        // Round device orientation to a multiple of 90
+        deviceOrientation = (deviceOrientation + 45) / 90 * 90
+
+        // Reverse device orientation for front-facing cameras
+        val facingFront =
+            c[CameraCharacteristics.LENS_FACING] === CameraCharacteristics.LENS_FACING_FRONT
+        if (facingFront) deviceOrientation = -deviceOrientation
+
+        // Calculate desired JPEG orientation relative to camera orientation to make
+        // the image upright relative to the device orientation
+        return (sensorOrientation + deviceOrientation + 360) % 360
```

Comment:
Cool :)

Reply:
I guess it is but can't say it's due to my genius. Plus it seems like it does not work. It might disappear during the Easter break.

Reply:
okay, too bad